dvc/objects/db/base.py

```diff
@@ -203,7 +203,10 @@ def check(
             obj.check(self, check_hash=check_hash)
         except ObjectFormatError:
             logger.warning("corrupted cache file '%s'.", obj.path_info)
-            self.fs.remove(obj.path_info)
+            try:
+                self.fs.remove(obj.path_info)
+            except FileNotFoundError:
+                pass
```

Comment:
Added this as a workaround for 

```
___________________________ test_get_hash_dirty_file ___________________________
2021-08-08T21:35:14.5084283Z [gw2] linux -- Python 3.6.14 /opt/hostedtoolcache/Python/3.6.14/x64/bin/python
2021-08-08T21:35:14.5085018Z 
2021-08-08T21:35:14.5086171Z self = <dvc.objects.db.reference.ReferenceObjectDB object at 0x7f05417ea518>
2021-08-08T21:35:14.5088268Z hash_info = HashInfo(name='md5', value='437b930db84b8079c2dd804a71936b5f', size=9, nfiles=None, obj_name=None)
2021-08-08T21:35:14.5089328Z check_hash = True
2021-08-08T21:35:14.5089863Z 
2021-08-08T21:35:14.5090685Z     def check(
2021-08-08T21:35:14.5091278Z         self,
2021-08-08T21:35:14.5091910Z         hash_info: "HashInfo",
2021-08-08T21:35:14.5092832Z         check_hash: bool = True,
2021-08-08T21:35:14.5093500Z     ):
2021-08-08T21:35:14.5094392Z         """Compare the given hash with the (corresponding) actual one if
2021-08-08T21:35:14.5095544Z         check_hash is specified, or just verify the existence of the cache
2021-08-08T21:35:14.5096748Z         files on the filesystem.
2021-08-08T21:35:14.5097384Z     
2021-08-08T21:35:14.5098423Z         - Use `State` as a cache for computed hashes
2021-08-08T21:35:14.5099601Z             + The entries are invalidated by taking into account the following:
2021-08-08T21:35:14.5101339Z                 * mtime
2021-08-08T21:35:14.5102217Z                 * inode
2021-08-08T21:35:14.5102735Z                 * size
2021-08-08T21:35:14.5103235Z                 * hash
2021-08-08T21:35:14.5103688Z     
2021-08-08T21:35:14.5105093Z         - Remove the file from cache if it doesn't match the actual hash
2021-08-08T21:35:14.5105858Z         """
2021-08-08T21:35:14.5106328Z     
2021-08-08T21:35:14.5106890Z         obj = self.get(hash_info)
2021-08-08T21:35:14.5107604Z         if self.is_protected(obj.path_info):
2021-08-08T21:35:14.5108861Z             logger.trace(  # type: ignore[attr-defined]
2021-08-08T21:35:14.5110099Z                 "Assuming '%s' is unchanged since it is read-only",
2021-08-08T21:35:14.5110986Z                 obj.path_info,
2021-08-08T21:35:14.5111545Z             )
2021-08-08T21:35:14.5112052Z             return
2021-08-08T21:35:14.5112553Z     
2021-08-08T21:35:14.5113040Z         try:
2021-08-08T21:35:14.5113896Z >           obj.check(self, check_hash=check_hash)
2021-08-08T21:35:14.5114384Z 
2021-08-08T21:35:14.5114991Z /home/runner/work/dvc/dvc/dvc/objects/db/base.py:203: 
2021-08-08T21:35:14.5115629Z _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
2021-08-08T21:35:14.5116160Z 
2021-08-08T21:35:14.5117297Z self = <dvc.objects.reference.ReferenceHashFile object at 0x7f05417ea5c0>
2021-08-08T21:35:14.5118902Z odb = <dvc.objects.db.reference.ReferenceObjectDB object at 0x7f05417ea518>
2021-08-08T21:35:14.5120483Z check_hash = True
2021-08-08T21:35:14.5120905Z 
2021-08-08T21:35:14.5121568Z     def check(self, odb: "ObjectDB", check_hash: bool = True):
2021-08-08T21:35:14.5122279Z         assert self.fs
2021-08-08T21:35:14.5122970Z         if not self.fs.exists(self.path_info):
2021-08-08T21:35:14.5123864Z             raise FileNotFoundError(
2021-08-08T21:35:14.5124732Z                 errno.ENOENT, os.strerror(errno.ENOENT), self.path_info
2021-08-08T21:35:14.5125463Z             )
2021-08-08T21:35:14.5126392Z         if not (self.mtime, self.size) == self._get_mtime_and_size():
2021-08-08T21:35:14.5127249Z >           raise ObjectFormatError(f"{self} is changed")
2021-08-08T21:35:14.5128889Z E           dvc.objects.errors.ObjectFormatError: ref object md5: 437b930db84b8079c2dd804a71936b5f -> file is changed
2021-08-08T21:35:14.5129894Z 
2021-08-08T21:35:14.5133457Z /home/runner/work/dvc/dvc/dvc/objects/reference.py:51: ObjectFormatError
2021-08-08T21:35:14.5133997Z 
2021-08-08T21:35:14.5134574Z During handling of the above exception, another exception occurred:
2021-08-08T21:35:14.5135058Z 
2021-08-08T21:35:14.5136182Z self = <fsspec.implementations.memory.MemoryFileSystem object at 0x7f05417ea6a0>
2021-08-08T21:35:14.5137550Z path = '/file'
2021-08-08T21:35:14.5137828Z 
2021-08-08T21:35:14.5138194Z     def _rm(self, path):
2021-08-08T21:35:14.5138786Z         path = self._strip_protocol(path)
2021-08-08T21:35:14.5139240Z         try:
2021-08-08T21:35:14.5139654Z >           del self.store[path]
2021-08-08T21:35:14.5140313Z E           KeyError: '/file'
2021-08-08T21:35:14.5140609Z 
2021-08-08T21:35:14.5141639Z /opt/hostedtoolcache/Python/3.6.14/x64/lib/python3.6/site-packages/fsspec/implementations/memory.py:207: KeyError
2021-08-08T21:35:14.5142337Z 
2021-08-08T21:35:14.5143010Z The above exception was the direct cause of the following exception:
2021-08-08T21:35:14.5143469Z 
2021-08-08T21:35:14.5144530Z tmp_dir = PosixTmpDir('/tmp/pytest-of-runner/pytest-0/popen-gw2/test_get_hash_dirty_file0')
2021-08-08T21:35:14.5146304Z dvc = Repo: '/tmp/pytest-of-runner/pytest-0/popen-gw2/test_get_hash_dirty_file0'
2021-08-08T21:35:14.5146871Z 
2021-08-08T21:35:14.5147317Z     def test_get_hash_dirty_file(tmp_dir, dvc):
2021-08-08T21:35:14.5148008Z         from dvc.objects import check
2021-08-08T21:35:14.5148699Z         from dvc.objects.errors import ObjectFormatError
2021-08-08T21:35:14.5149467Z         from dvc.objects.stage import get_file_hash
2021-08-08T21:35:14.5149975Z     
2021-08-08T21:35:14.5150374Z         tmp_dir.dvc_gen("file", "file")
2021-08-08T21:35:14.5151060Z         file_hash_info = HashInfo("md5", "8c7dd922ad47494fc02c388e12c00eac")
2021-08-08T21:35:14.5151644Z     
2021-08-08T21:35:14.5152079Z         (tmp_dir / "file").write_text("something")
2021-08-08T21:35:14.5152783Z         something_hash_info = HashInfo("md5", "437b930db84b8079c2dd804a71936b5f")
2021-08-08T21:35:14.5153355Z     
2021-08-08T21:35:14.5153826Z         clean_staging()
2021-08-08T21:35:14.5154181Z     
2021-08-08T21:35:14.5154577Z         # file is modified in workspace
2021-08-08T21:35:14.5155339Z         # get_file_hash(file) should return workspace hash, not DVC cached hash
2021-08-08T21:35:14.5156020Z         fs = RepoFileSystem(repo=dvc)
2021-08-08T21:35:14.5156649Z         assert fs.info(PathInfo(tmp_dir) / "file").get("md5") is None
2021-08-08T21:35:14.5157394Z         staging, obj = stage(dvc.odb.local, PathInfo(tmp_dir) / "file", fs, "md5")
2021-08-08T21:35:14.5158104Z         assert obj.hash_info == something_hash_info
2021-08-08T21:35:14.5158627Z         check(staging, obj)
2021-08-08T21:35:14.5159129Z     
2021-08-08T21:35:14.5159537Z         # file is removed in workspace
2021-08-08T21:35:14.5160173Z         # any staged object referring to modified workspace obj is now invalid
2021-08-08T21:35:14.5160790Z         (tmp_dir / "file").unlink()
2021-08-08T21:35:14.5161499Z         with pytest.raises(ObjectFormatError):
2021-08-08T21:35:14.5162213Z >           check(staging, obj)
2021-08-08T21:35:14.5162516Z 
2021-08-08T21:35:14.5163021Z /home/runner/work/dvc/dvc/tests/unit/fs/test_repo.py:607: 
2021-08-08T21:35:14.5163572Z _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
2021-08-08T21:35:14.5164142Z /home/runner/work/dvc/dvc/dvc/objects/__init__.py:20: in check
2021-08-08T21:35:14.5164767Z     odb.check(obj.hash_info, **kwargs)
2021-08-08T21:35:14.5165396Z /home/runner/work/dvc/dvc/dvc/objects/db/base.py:206: in check
2021-08-08T21:35:14.5166044Z     self.fs.remove(obj.path_info)
2021-08-08T21:35:14.5166896Z /home/runner/work/dvc/dvc/dvc/fs/fsspec_wrapper.py:117: in remove
2021-08-08T21:35:14.5168100Z     self.fs.rm_file(self._with_bucket(path_info))
2021-08-08T21:35:14.5169332Z /opt/hostedtoolcache/Python/3.6.14/x64/lib/python3.6/site-packages/fsspec/spec.py:886: in rm_file
2021-08-08T21:35:14.5170090Z     self._rm(path)
2021-08-08T21:35:14.5170511Z _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
2021-08-08T21:35:14.5170908Z 
2021-08-08T21:35:14.5171908Z self = <fsspec.implementations.memory.MemoryFileSystem object at 0x7f05417ea6a0>
2021-08-08T21:35:14.5173235Z path = '/file'
2021-08-08T21:35:14.5173509Z 
2021-08-08T21:35:14.5173896Z     def _rm(self, path):
2021-08-08T21:35:14.5174570Z         path = self._strip_protocol(path)
2021-08-08T21:35:14.5175047Z         try:
2021-08-08T21:35:14.5175482Z             del self.store[path]
2021-08-08T21:35:14.5176074Z         except KeyError as e:
2021-08-08T21:35:14.5176639Z >           raise FileNotFoundError from e
2021-08-08T21:35:14.5177228Z E           FileNotFoundError
2021-08-08T21:35:14.5177564Z 
2021-08-08T21:35:14.5178856Z /opt/hostedtoolcache/Python/3.6.14/x64/lib/python3.6/site-packages/fsspec/implementations/memory.py:209: FileNotFoundError
```

caused by `_obj_cache` diverging with an actual memfs situation after `cleanup_staging`. Seems useful though, but for cases where two odb instances might be checking the same object file and removing it. But regarding the actual cause here, looks like `cleanup_staging` doesn't have an odb instance to cleanup the dict cache from, so maybe we need to also make it global (like memfs) or maybe we just need to scope out staging db scope (e.g. maybe have one instance per repo, as we've discussed before). :thinking: